### DISCUSSION TITLE
Query hints on CTEs

### DISCUSSION THREAD

**Discussion by sonthonaxrk** [sqlalchemy/sqlalchemy#7785] [Usage Questions]
Hey there :) 

Simple question this time.  I'm trying to get pg_hint_plan style hint on a CTE object.  However, I can't actually figure out where to put the `prefix_with` clause.  I need to get the query hint to appear before the WITH clause, not inside the statement.

This is what I want

```sql
/*+ query hint */
WITH cte AS
(SELECT test.a AS a, test.b AS b
FROM test)
 SELECT test.a, test.b
FROM test, cte
```

But this is what I get

```sql
WITH cte AS /*+ query hint */
(SELECT test.a AS a, test.b AS b
FROM test)
 SELECT test.a, test.b
FROM test, cte
```

Example code: 

```python
from sqlalchemy import *

t1 = table('test', Column('a'), Column('b'))


out = str(
    select(t1)
    .select_from(
        select(t1).cte('cte').prefix_with('/*+ query hint */')
    )
)

assert out == '/*+ query hint */ WITH cte AS \n(SELECT test.a AS a, test.b AS b \nFROM test)\n SELECT test.a, test.b \nFROM test, cte'
```

Any ideas?

**Comment by zzzeek** [comment]
well that's probably a bug, that's a weird place for it to put that hint.   the only workaround for now is to use plain text or to use a `@compiles` event, something like that.  I dont have a clean solution other than fix the bug

**Reply by zzzeek** [reply]
im not sure that's the correct use of prefix_with.

**Reply by zzzeek** [reply]
can you show me what database this is and where this syntax is documented?

**Reply by zzzeek** [reply]
yah for CTEs, prefixes means a specific PG syntax like this:

```
 WITH regional_sales AS NOT MATERIALIZED 
 (SELECT "order"."order" AS "order" 
 FROM "order") SELECT "order"."order" FROM "order", 
 regional_sales WHERE "order"."order" > regional_sales."order"
```

looks like your syntax has to do with MySQL.  that would be a use case that needs to be somehow implemented.  likely as a hint, maybe a with_hint() method on cte().

**Reply by sonthonaxrk** [reply]
It's https://github.com/ossc-db/pg_hint_plan.  It's the a query planner hint API for Postgres and it's available on RDS and Aurora, so it's by no means niche, 

It's not so much a syntax as just

```
/*+ hint */
STATEMENT ...
```

**Reply by sonthonaxrk** [reply]
```
from sqlalchemy import *
from sqlalchemy.ext.compiler import compiles

from sqlalchemy.sql.expression import Executable, ClauseElement


class PGHint(Executable, ClauseElement):
    inherit_cache = False

    def __init__(self, hint, select):
        self.hint = hint
        self.select = select


@compiles(PGHint)
def visit_pg_hint(element, compiler, **kw):
    return "/*+ %s */\n %s" % (
        element.hint,
        compiler.process(element.select, **kw)
    )


t1 = table('test', Column('a'), Column('b'))
cte = select(t1).cte('cte')
sel = select(t1).select_from(cte)
cte_comment = PGHint("here", sel)
print(cte_comment)
```

This is probably the solution.

**Reply by sonthonaxrk** [reply]
Actually, no this breaks the ORM query.  Execute just returns raw rows now, without the ORM objects mapped - weird.

Is there any metadata that I need to pass through?

**Reply by sonthonaxrk** [reply]
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/sql/compiler.py#L2937

Yeah, I don't think I can really fix the prefixing (or to use a better term, make this kind of prefixing work) without overriding the whole `visit_cte` function.

**Reply by zzzeek** [reply]
> It's https://github.com/ossc-db/pg_hint_plan. It's the a query planner hint API for Postgres and it's available on RDS and Aurora, so it's by no means niche,
> 
> It's not so much a syntax as just
> 
> ```
> /*+ hint */
> STATEMENT ...
> ```

is this part of PostgreSQL?     or this is some third party syntax?

> Yeah, I don't think I can really fix the prefixing (or to use a better term, make this kind of prefixing work) without overriding the whole visit_cte function.

but you want it to go *before* the WITH entirely.  You would get the whole compiler output from visit_select(), then stick the hint thing ahead of it.

So far this is looking a lot like "recipe" to me since it seems to be a third party thing and it looks a whole lot like MySQL's syntax.

**Reply by sonthonaxrk** [reply]
It's third party, but it's the only query hinting modules that Postgres has, and it's ubiquitous on the cloud providers.

**Comment by zzzeek** [comment]
Here's a recipe for this.  this seems to be something third party so we'd look to make this kind of recipe in the wiki, or similar.   UserDefinedOption is public API but as this option affects the compilation, we need to add cache directives to prevent it from caching.

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import String
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session
from sqlalchemy.orm import UserDefinedOption
from sqlalchemy.sql.base import HasCacheKey
from sqlalchemy.sql.expression import Select
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    data = Column(String)


@compiles(Select)
def _add_osscdbhints(element, compiler, **kw):
    text = compiler.visit_select(element, **kw)
    for opt in element._with_options:
        if isinstance(opt, osscdbhint):
            text = opt.text + " " + text
    return text

class osscdbhint(HasCacheKey, UserDefinedOption):
    inherit_cache = False

    def __init__(self, text):
        self.text = text


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

stmt =(
    select(A)
    .select_from(
        select(A).cte('cte')
    ).options(osscdbhint("/*+ hint */"))
)
s.execute(stmt)

```

**Reply by sonthonaxrk** [reply]
Thank you sir.

**Reply by sonthonaxrk** [reply]
Would you be able to give me a rundown of what I need to do to make that cacheable?  Obviously something wrapping everything else is the worst place not to cache.

**Reply by zzzeek** [reply]
this will do it.  it's just not API we've published yet as caching is still very new.

```py
class osscdbhint(HasCacheKey, UserDefinedOption):

    def _gen_cache_key(self, anon_map, bindparams):
        return (osscdbhint, self.text)

    def __init__(self, text):
        self.text = text
```
